File: js/4386.js
const fs = require('fs');

const inputs = fs
  .readFileSync('example.txt')
  .toString()
  .split('\n')
  .map((e) => e.split(' ').map(Number));

let idx = 0;
const getLine = () => inputs[idx++];

let [n] = getLine();

const coords = [];
for (let i = 0; i < n; i++) {
  coords.push(getLine());
}

const edges = [];

for (let i = 0; i < coords.length; i++) {
  const [x1, y1] = coords[i];
  for (let j = i + 1; j < coords.length; j++) {
    const [x2, y2] = coords[j];
    const weight =
      Math.floor(Math.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2) * 100) / 100;
    edges.push([i, j, weight]);
  }
}

edges.sort((a, b) => a[2] - b[2]);

const parent = Array(n)
  .fill()
  .map((_, i) => i);

let answer = 0;
for (let i = 0; i < edges.length; i++) {
  const [a, b, weight] = edges[i];

  if (!sameParent(a, b)) {
    unionParent(a, b);
    answer += weight;
  }
}

console.log(answer);

function getParent(x) {
  if (x === parent[x]) return x;
  else return (parent[x] = getParent(parent[x]));
}

function unionParent(a, b) {
  a = getParent(a);
  b = getParent(b);
  if (a < b) parent[b] = a;
  else parent[a] = b;
}

function sameParent(a, b) {
  a = getParent(a);
  b = getParent(b);
  return a === b;
}

File: c++/4386.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <vector>

using namespace std;

double coords[100][2];
vector<pair<double, pair<double, double>>> edges;
int parent[10001];

int getParent(int x) {
  if (x == parent[x])
    return x;
  else
    return parent[x] = getParent(parent[x]);
}

void unionParnet(int a, int b) {
  a = getParent(a);
  b = getParent(b);
  if (a < b)
    parent[b] = a;
  else
    parent[a] = b;
}

bool sameParent(int a, int b) {
  a = getParent(a);
  b = getParent(b);
  return a == b;
}

int main() {
  freopen("example.txt", "r", stdin);

  ios_base::sync_with_stdio(0);
  cin.tie(0);
  int n;
  cin >> n;

  for (int i = 0; i < n; i++) {
    cin >> coords[i][0] >> coords[i][1];
  }

  for (int i = 0; i < n - 1; i++) {
    for (int j = i + 1; j < n; j++) {
      double x1 = coords[i][0];
      double y1 = coords[i][1];
      double x2 = coords[j][0];
      double y2 = coords[j][1];

      double weight = sqrt((pow((x2 - x1), 2) + pow((y2 - y1), 2)));
      edges.push_back({weight, {i, j}});
    }
  }
  for (int i = 0; i < n; i++) {
    parent[i] = i;
  }

  sort(edges.begin(), edges.end());

  double answer = 0;
  for (int i = 0; i < edges.size(); i++) {
    double weight = edges[i].first;
    int a = edges[i].second.first;
    int b = edges[i].second.second;

    if (!sameParent(a, b)) {
      unionParnet(a, b);
      answer += weight;
    }
  }

  cout << answer;

  return 0;
}